binary_search_tree/tree.py

```diff
@@ -14,45 +16,148 @@ class Tree:
+                elif key > current.key:
+                    if current.right == None:
+                        current.right = tree
+                    current = current.right 
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        
+        if self.root == None:
+            return None
+
+        current = self.root
+
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def inorder(self):
-        pass
+        tree_nodes = []
+
+        if self.root == None:
+            return tree_nodes
+        else:
+            return self.inorder_helper(self.root, tree_nodes)
+
+    def inorder_helper(self, current, tree_nodes):
+        if current == None:
+            return None
+
+        self.inorder_helper(current.left, tree_nodes)
+        
+        tree_nodes.append(
+                {
+                    "key": current.key, 
+                    "value": current.value
+                }
+            )
 
-    # Time Complexity: 
-    # Space Complexity:     
+        self.inorder_helper(current.right, tree_nodes)
+
+        return tree_nodes
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)     
     def preorder(self):
-        pass
+        tree_nodes = []
+
+        if self.root == None: 
+            return tree_nodes 
+        else:
+            return self.preorder_helper(self.root, tree_nodes)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def preorder_helper(self, current, tree_nodes):
+        if current == None: 
+            return None 
+
+        tree_nodes.append(
+            {
+                "key": current.key,
+                "value": current.value
+            }
+        )
+
+        self.preorder_helper(current.left, tree_nodes)
+
+        self.preorder_helper(current.right, tree_nodes)
+
+        return tree_nodes
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)     
     def postorder(self):
-        pass
+        tree_nodes = []
+
+
+        return self.postorder_helper(self.root, tree_nodes)
+    
+    def postorder_helper(self, current, tree_nodes):
+        if current == None:
+            return tree_nodes
+
+        self.postorder_helper(current.left, tree_nodes)
+
+        self.postorder_helper(current.right, tree_nodes)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        tree_nodes.append(
+            {
+                "key": current.key,
+                "value": current.value
+            }
+        )
+
+        return tree_nodes
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def height(self):
```

Comment:
👍 Because of the recursion the space complexity is O(h) where h is the height of the tree (which will be log n if the tree is balanced and it can approach n if the tree is very unbalanced).